azurerm/internal/services/azurestackhci/hci_cluster_resource.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmHCICluster() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmHCIClusterCreate,
+		Read:   resourceArmHCIClusterRead,
+		Update: resourceArmHCIClusterUpdate,
+		Delete: resourceArmHCIClusterDelete,
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Importer: azSchema.ValidateResourceIDPriorToImport(func(id string) error {
+			_, err := parse.HCIClusterID(id)
+			return err
+		}),
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.HCIClusterName,
+			},
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"location": azure.SchemaLocation(),
+
+			"client_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.IsUUID,
+			},
+
+			"tenant_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.IsUUID,
+			},
+
+			"tags": tags.Schema(),
+		},
+	}
+}
+func resourceArmHCIClusterCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).AzureStackHCI.ClusterClient
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	existing, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		if !utils.ResponseWasNotFound(existing.Response) {
+			return fmt.Errorf("checking for present of existing HCI Cluster %q (Resource Group %q): %+v", name, resourceGroup, err)
+		}
+	}
+
+	if existing.ID != nil && *existing.ID != "" {
+		return tf.ImportAsExistsError("azurerm_hci_cluster", *existing.ID)
+	}
+
+	cluster := azurestackhci.Cluster{
+		Location: utils.String(location.Normalize(d.Get("location").(string))),
+		ClusterProperties: &azurestackhci.ClusterProperties{
+			AadClientID: utils.String(d.Get("client_id").(string)),
+			AadTenantID: utils.String(d.Get("tenant_id").(string)),
+		},
+		Tags: tags.Expand(d.Get("tags").(map[string]interface{})),
+	}
+
+	if _, err := client.Create(ctx, resourceGroup, name, cluster); err != nil {
+		return fmt.Errorf("creating HCI Cluster %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		return fmt.Errorf("retrieving HCI Cluster %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	if resp.ID == nil || *resp.ID == "" {
+		return fmt.Errorf("empty or nil ID returned for HCI Cluster %q (Resource Group %q) ID", name, resourceGroup)
+	}
+
+	d.SetId(*resp.ID)
```

Comment:
Can we construct an ID here and set the result of `id.ID()` back as the argument of `d.SetId()`?

Reply:
Could you share me an example PR? Thanks.